File: tree/102.c
/**
 * key在于递归时带一个depth 可以保证每个level的准确性
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *columnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
void traverse(int*** s, int** columnSizes, int* returnSize, struct TreeNode* t, int depth);

int** levelOrder(struct TreeNode* root, int** columnSizes, int* returnSize) {
    int*** s = (int***)malloc(sizeof(int**));
    *s = NULL;
    if(root==NULL) return *s;
    *returnSize = 0;
    traverse(s, columnSizes, returnSize, root, 0);
    return *s;
}

void traverse(int*** s, int**columnSizes, int* returnSize, struct TreeNode* t, int depth)
{
    if(t==NULL) return;
    if(*returnSize < depth + 1)
    {
        *returnSize = depth + 1;
        *s = (int**)realloc(*s, (depth+1)*sizeof(int*));
        (*s)[depth] = NULL;
        *columnSizes = (int*)realloc(*columnSizes, (depth+1)*sizeof(int*));
        (*columnSizes)[depth] = 0;
    }
    (*s)[depth] = (int*)realloc((*s)[depth], ((*columnSizes)[depth]+1)*sizeof(int));
    (*s)[depth][(*columnSizes)[depth]] = t->val;
    ++(*columnSizes)[depth];
    traverse(s, columnSizes, returnSize, t->left, depth+1);
    traverse(s, columnSizes, returnSize, t->right, depth+1);
}

File: tree/102.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type Item struct{
    level int 
    t *TreeNode
}

func Slicedelete(s []*Item, index int) []*Item {
    return append([]*Item{}, s[index+1:]...)
}
func levelOrder(root *TreeNode) [][]int {
    ret := make([][]int, 0)
    list := make([]*Item, 0)
    level := 0
    if root == nil {
        return nil
    }
    list = append(list, &Item{0, root})
    for len(list) > 0 {
        temp := make([]int, 0)
        for len(list) > 0 && list[0].level == level {
            root := list[0].t
            list = list[1:]
            
            temp = append(temp, root.Val)
            if root.Left != nil {
                list = append(list, &Item{level+1, root.Left})
            }
            if root.Right != nil {
                list = append(list, &Item{level+1, root.Right})
            }
        }
        ret = append(ret, temp)
        level++
    }
    return ret
}
